src/jit/lower.cpp

```diff
@@ -491,7 +491,63 @@ void Lowering::DecomposeNode(GenTreePtr* pTree, Compiler::fgWalkData* data)
         }
         break;
     case GT_NEG:
-        NYI("GT_NEG of TYP_LONG");
+        {
+            GenTree* op1 = tree->gtGetOp1();
+            noway_assert(op1->OperGet() == GT_LONG);
+
+            // TODO-Cleanup: Identical code exists in DecomposeStoreInd,
+            // it should be moved to a function similar to CreateTemporary
+            GenTreeStmt* newStmt = comp->fgInsertEmbeddedFormTemp(&op1->gtOp.gtOp1);
+            if (newStmt->gtStmtIsTopLevel())
+            {
+                for (GenTreeStmt* nextEmbeddedStmt = newStmt->gtStmtNextIfEmbedded();
+                     nextEmbeddedStmt != nullptr;
+                     nextEmbeddedStmt = nextEmbeddedStmt->gtStmt.gtStmtNextIfEmbedded())
+                {
+                    comp->compCurStmt = nextEmbeddedStmt;
+                    comp->fgWalkTreePost(&nextEmbeddedStmt->gtStmt.gtStmtExpr, Lowering::DecompNodeHelper, this, true);
+                }
+
+                comp->compCurStmt = curStmt;
+            }
+
+            newStmt = comp->fgInsertEmbeddedFormTemp(&op1->gtOp.gtOp2);
+            assert(!newStmt->gtStmtIsTopLevel());  
+
+            GenTree* loOp1 = op1->gtGetOp1();
+            GenTree* hiOp1 = op1->gtGetOp2();
+
+            //
+            // Change the existing GT_NEG node so it computes the high part
+            // and insert the low part computation right before it.
+            //
+
+            GenTree* loNeg = comp->gtNewOperNode(GT_NEG, TYP_INT, loOp1);
+            loNeg->CopyCosts(tree);
+            loNeg->gtFlags = (tree->gtFlags & GTF_LIVENESS_MASK);
```

Comment:
Yeah, `GTF_LIVENESS_MASK` only applies to lclVar nodes (it's too bad we don't have better debugging modes to detect reading/writing bits you're not supposed to touch, based on node type). It's probably ok to copy costs, although as you point out, it might not matter.

---
In reply to: [70025072](https://github.com/dotnet/coreclr/pull/5966#discussion_r70025072) [](ancestors = 70025072)